### PR TITLE
GRIDEDIT-795: Create conda env in docker config

### PR DESCRIPTION
**Pull request by ahmad-el-sayed**

### CODE REVIEW

**Comment by matthijs-deltares** [comment]
Hi Ahmed,

You should try to run your *.sh scripts through https://www.shellcheck.net/. I already notice a couple things:
- add #!/bin/sh to the first line of your script ->https://www.shellcheck.net/wiki/SC2148
- Double quote to prevent globbing and word splitting e.g. "${DEVTOOLSET}" instead of ${DEVTOOLSET}
- use cd ... || exit for in case cd fails due to an incorrect path https://www.shellcheck.net/wiki/SC2164

As we discussed in person you can run this command on the scripts to give them executing permissions:
git update-index --chmod=+x scripts/*.sh -> https://stackoverflow.com/questions/21691202/how-to-create-file-execute-mode-permissions-in-git-on-windows
Then you can remove the teamcity build step

File: scripts/Dockerfile [hunk 1]
```diff
@@ -2,18 +2,28 @@
...
 
-# add gcc compiler, devtoolset-11
-ENV devtoolset=devtoolset-11
-RUN ./compile_deps.sh
+# versions of deps (must be set pripor to running install_deps.sh)
+ENV DEVTOOLSET=devtoolset-11
+ENV CMAKE_VERSION="3.23.1"
+ENV BOOST_VERSION="1.81.0"
+ENV MINICONDA3_VERSION="py38_4.10.3"
+ENV PYTHON_VERSION="3.8"
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply scripts/Dockerfile:15]: This var in particular is used in build_wheel.sh (see CMD). exporting it in install_deps.sh will not keep in the env until CMD executes.

File: scripts/Dockerfile [hunk 2]
```diff
@@ -2,18 +2,28 @@
 # By using an old linux operating system, dynamic libraries should be compatible with most linux os
 
 FROM quay.io/pypa/manylinux2014_x86_64
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply scripts/Dockerfile:4]: Will select a stable version

File: scripts/Dockerfile [hunk 3]
```diff
@@ -2,18 +2,28 @@
...
+
 WORKDIR /root
 
 ADD . .
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply scripts/Dockerfile:8]: done

File: scripts/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply scripts/Dockerfile:4]: Chose latest available tag, [2023-11-20-745eb52](https://quay.io/repository/pypa/manylinux2014_x86_64?tab=history&tag=latest) at the time of writing.

File: scripts/install_deps.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+
+error() {
+  # store last exit code before invoking any other command
+  local EXIT_CODE=$?
+  # print error message
+  echo $(basename "$0"): $1
```
Comment (matthijs-deltares) [inline scripts/install_deps.sh:7]:
```suggestion
   echo $(basename "$0"): "$1"
```

File: scripts/install_deps.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  exit $EXIT_CODE
+}
+
+# add development tools
+yum -y install git make wget which bzip2 netcdf centos-release-scl
+yum -y install ${DEVTOOLSET}
```
Comment (matthijs-deltares) [inline scripts/install_deps.sh:13]:
```suggestion
yum -y install "${DEVTOOLSET}"
```

File: scripts/install_deps.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+scl enable ${DEVTOOLSET} bash
```
Comment (matthijs-deltares) [inline scripts/install_deps.sh:14]: scl enable "${DEVTOOLSET}" bash

File: scripts/install_deps.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+# enter root
+(
+  cd /root  || error "Could not change the directory to root"
+
+  # install CMake
+  CMAKE_SCRIPT=cmake-${CMAKE_VERSION}-linux-x86_64.sh
+  wget https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION}/${CMAKE_SCRIPT} \
```
Comment (matthijs-deltares) [inline scripts/install_deps.sh:22]:
```suggestion
  wget https://github.com/Kitware/CMake/releases/download/v"${CMAKE_VERSION}"/"${CMAKE_SCRIPT}" \
```

File: scripts/install_deps.sh [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    || error "[cmake] ${CMAKE_SCRIPT} download failed"
+  CMAKE_INSTALL_PREFIX=/opt/cmake
+  mkdir ${CMAKE_INSTALL_PREFIX} || error "[cmake] Creation of ${CMAKE_INSTALL_PREFIX} failed"
+  chmod +x ${CMAKE_SCRIPT} || error "[cmake] Changing the permissions of ${CMAKE_SCRIPT} failed"
```
Comment (matthijs-deltares) [inline scripts/install_deps.sh:26]:
```suggestion
  chmod +x "${CMAKE_SCRIPT}" || error "[cmake] Changing the permissions of ${CMAKE_SCRIPT} failed"
```

File: scripts/build_wheel.sh
Comment (matthijs-deltares) [inline scripts/build_wheel.sh:1]: I still see a lot of variables defined without double quotes. I placed comments on a few. 
You can check your scripts easily on https://www.shellcheck.net/

File: scripts/build_wheel.sh [hunk 9]
```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
+
+DOCKER_CONDA_ENV="docker_conda_env"
+
+# create conda env and activate it
+conda create -y --force -n ${DOCKER_CONDA_ENV} python=${PYTHON_VERSION} pip
+source activate ${DOCKER_CONDA_ENV}
```
Comment (matthijs-deltares) [inline scripts/build_wheel.sh:7]: use dot instead of source
```suggestion
. activate ${DOCKER_CONDA_ENV}
```

File: scripts/build_wheel.sh [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  # deactivate and remove conda env
+  conda deactivate
+  conda env remove -n ${DOCKER_CONDA_ENV}
+}
+
+error() {
+  # store last exit code before invoking any other command
+  local EXIT_CODE=$?
+  # print error message
+  echo $(basename "$0"): $1
```
Comment (matthijs-deltares) [inline scripts/build_wheel.sh:19]:
```suggestion
  echo $(basename "$0"): "$1"
```

**Comment by matthijs-deltares** [comment]
In general I would put your scripts in https://www.shellcheck.net/ it gives great suggestions which you often can apply right away. If you fix the warnings then I'm fine with merging :)

File: scripts/build_wheel.sh
Comment (ahmad-el-sayed) [reply scripts/build_wheel.sh:1]: all instances fixed

**Pull request merged**